Fix issue #309

f90 opened the pull request:
Fix for issue #309 

Other changes:
Create function for getting all guild members and filtering according to on/offline and same-zone criteria
Better variable naming and comments

However, this is **completely untested** and probably **doesnt work yet** as I don't know how to test within WoW when I already have Monolith installed for my Guild and don't want to mess anything up with my own custom version. Can someone give advice on how to test or do it on their end?

f90 committed:
Fix issue #309
Create function for getting all guild members and filtering according to on/offline and same-zone criteria
Better variable naming and comments

lantisnt commented:
Change all tabs back to  2 spaces, as your changes are completly not visible :)

lantisnt commented:
I have separate guild for that on one of my alts. One thing: please change tabs to spaces when editing code :)

f90 commented:
Does it automatically create separate DKP tables and all if you are in a different guild on a different character on the same server? So that it doesnt interfere?

For the indentation I am a bit confused, if I look at other code in this repo i see tabs everywhere - whats the standard for this repo? Tabs? 3 spaces?

lantisnt commented:
you need to swap them manually, i just dont care and wipe mine, ill get a sync from an officer anytime

I didnt find any rules for the style and theres many different in a code so i assume standard lua (2tabs per ident) is in place, right? or i am thinking wrong?

f90 commented:
OK i am going to test it on my end by backing up the WTF folder beforehand and then replacing all the Monolith DKP files including savedvariables stuff.

With indentation I have no idea, I didn't do Lua programming before!

One shortcoming of this implementation by the way is that I couldn't find a way to check online/zone status for non-guild members on the standby list, so those will not be awarded DKP if one activates the same-zone OR online-only requirements, otherwise they will be! Is that fine? Or is there some way to check this for non-guild members?

f90 committed:
Fix indentation
Fix parenthesis error

f90 committed:
Local before global variable

f90 committed:
Local before global variable

f90 committed:
More code optimisations

Commit:
Code cleanup on how DKP is mass-awarded to raids and standby list (put all into AwardDKP.lua)
Fix with auto-award (boss-kill award) using incstandby option not autoincstandby to determine whether standby get DKP

f90 commented:
OK i personally tested the code now. I significantly cleaned up the code around DKP awarding (e.g. only one function to award raid(+standby) some DKP, and reusing that function for both time-based and boss-kill DKP). And now standbys are properly handled (excluded when they are offline or in the wrong zone depending on options set). From my end this is ready to merge.

lantisnt commented on Modules/AwardDKP.lua:
```diff
@@ -0,0 +1,105 @@
+    -- Get list of guild members, optionally filtered for only online members and/or members currently in zone currZone
+    -- Returns table playerTable where playerTable[playerName] is not nil iff playerName is eligible
+    local playerTable = {}
+    for playerIndex = 1, GetNumGuildMembers() do
```
Make sure `GuildRoster()` is called before this call -> https://wowwiki.fandom.com/wiki/API_GuildRoster

f90 replied:
```suggestion
    GuildRoster()
    for playerIndex = 1, GetNumGuildMembers() do
```

lantisnt commented on Modules/AwardDKP.lua:
```diff
@@ -0,0 +1,105 @@
+
+function MonDKP:AwardPlayer(name, amount)
+	local search = MonDKP:Table_Search(MonDKP_DKPTable, name, "player")
+	local path;
```
`path` is a bit confisuing name in this context :)

f90 replied:
Changed to ```player``` - btw this was old code I just put to a different location, not by me ;)

lantisnt commented on Modules/AwardDKP.lua:
```diff
@@ -0,0 +1,105 @@
+    local curTime = time();
+    local curOfficer = UnitName("player")
+
+    if MonDKP:CheckRaidLeader() then -- only allows raid leader to disseminate DKP
```
Disagree. This limits functionality. It should be officer from whitelist. RL doesn't need to be the one responsible for DKP. For us ML does it.

f90 replied:
For boss-kill DKP ("autoaward") this check is in place right now, see
https://github.com/Roeshambo/MonolithDKP/blob/0f5f77db5c19c8af192a0607f8d8da9cfb32cc19/Modules/AutoAward.lua

For time-based DKP (RaidTimer.lua), its also only given as raid-leader:
https://github.com/Roeshambo/MonolithDKP/blob/0f5f77db5c19c8af192a0607f8d8da9cfb32cc19/Modules/RaidTimer.lua#L119

So basically nothing is changed with this pull request in this regard. Why it is like that in the first place, i am not sure but @Roeshambo should know?

We could think about changing it in the future though. But it might create problems with keeping the DKP tables in sync if multiple officers assign DKP instead of one of them that then keeps syncing changes to them.

lantisnt commented on Modules/AwardDKP.lua:
```diff
@@ -0,0 +1,105 @@
+local MonDKP = core.MonDKP;
+local L = core.L;
+
+local function GetEligibleGuildMembers(onlineOnly, sameZone, currZone)
```
Is `onlineOnly` and `sameZone` configurable through UI?

f90 replied:
Yes, and it is automatically set based on those options since in the same file i do
```
local standbyEligible = GetEligibleGuildMembers(MonDKP_DB.modes.OnlineOnly, MonDKP_DB.modes.SameZoneOnly, GetRealZoneText())
```
so the list of eligible guild members is filtered based on the current options.

lantisnt reviewed (changes requested):
Have you tested if it's backward compatible? This is a big rework (though kudos for it).

f90 committed:
Change path to player in AwardPlayer

f90 commented:
I tested it with another player that has officer rights with the latest public release version of the addon, and it seems like he sees different DKP in the table. I think the officers would have to update, but everyone else should be fine as DKP table changes get synced to them by the officer(s).

f90 committed:
Update Modules/AwardDKP.lua

f90 commented:
Can this pull request be confirmed/looked at please, @Roeshambo ?
Also I was wondering when the latest improvements will be packaged into a new release. The last release is already very long ago (December?)
